pylsp/_utils.py

```diff
@@ -144,17 +146,72 @@ def _merge_dicts_(a, b):
     return dict(_merge_dicts_(dict_a, dict_b))
 
 
-def format_docstring(contents):
-    """Python doc strings come in a number of formats, but LSP wants markdown.
-
-    Until we can find a fast enough way of discovering and parsing each format,
-    we can do a little better by at least preserving indentation.
+def escape_markdown(contents: str) -> str:
+    """
+    Try to make the plain text string work nicely when displayed in Markdown environment.
     """
+    # escape markdown syntax
+    contents = re.sub(r'([\\*_#[\]])', r'\\\1', contents)
+    # preserve white space characters
     contents = contents.replace('\t', '\u00A0' * 4)
     contents = contents.replace('  ', '\u00A0' * 2)
     return contents
 
 
+def wrap_signature(signature):
+    return '```python\n' + signature + '\n```\n'
+
+
+SERVER_SUPPORTED_MARKUP_KINDS = {'markdown', 'plaintext'}
+
+
+def choose_markup_kind(client_supported_markup_kinds: List[str]):
+    """Choose a markup kind supported by both client and the server,
+
+    giving a priority to the markup kinds provided earlier on the client preference list"""
+    for kind in client_supported_markup_kinds:
+        if kind in SERVER_SUPPORTED_MARKUP_KINDS:
+            return kind
+    return 'markdown'
+
+
+def format_docstring(contents: str, markup_kind: str, signatures: Optional[List[str]] = None):
+    """Transform the provided docstring into a MarkupContent object.
+
+    If `markup_kind` is 'markdown' the docstring will get converted to
+    markdown representation using `docstring-to-markdown`; if it is
+    `plaintext`, it will be returned as plain text.
+    """
+    if not isinstance(contents, str):
+        contents = ''
+
+    if markup_kind == 'markdown':
+        try:
+            value = docstring_to_markdown.convert(contents)
+            return {
+                'kind': 'markdown',
+                'value': value
+            }
+        except docstring_to_markdown.UnknownFormatError:
+            # try to escape the Markdown syntax instead:
+            value = escape_markdown(contents)
+
+        if signatures:
+            value = wrap_signature('\n'.join(signatures)) + '\n\n' + value
+
+        return {
+            'kind': 'markdown',
+            'value': value
+        }
+    value = contents
+    if signatures:
+        value = '\n'.join(signatures) + '\n\n' + value
+    return {
+        'kind': 'plaintext',
+        'value': value
```

Comment:
I think that returning the raw docstring value instead of what the previous definition of `format_docstring` did could break existing clients (e.g. us in Spyder). So I'd prefer to keep that small formatting here.

Reply:
It looks like Spyder already supports it:
- for completions since v5.0 (introduced 2 years in https://github.com/spyder-ide/spyder/pull/14314):
https://github.com/spyder-ide/spyder/blob/c0f446e0b0d46ccc9a99ee99163d6c668314a6bd/spyder/plugins/editor/widgets/completion.py#L503-L504
- for hover, forever: https://github.com/spyder-ide/spyder/blob/01d9195b1a33d02f42faf3c8e0047412966a24ce/spyder/plugins/completion/providers/languageserver/providers/document.py#L157-L166

But of course I can be wrong - I don't know your code base.

Also, the LSP spec says that this is non-optional: clients must handle a response like this, and plain strings are deprecated.

I can see a case for returning plain string performance-wise, but if we want to be future proof, keeping it as proposed would ease migration to LSP 4.0 if the deprecated string response gets dropped.

What do you think?

Reply:
> It looks like Spyder already supports it

Right, it does.

Perhaps I didn't explain myself correctly, but it could also be the case that I don't understand very well what's happening here.

In any case, my understanding is that right now clients are receiving docstrings as plain text but with a small formatting done by `format_docstring`, which preserves white space characters. If that's correct, I'd like to keep that small enhancement on this line (of course, we could move that code to Spyder, but I don't know if other clients depend on it).

Reply:
> In any case, my understanding is that right now clients are receiving docstrings as plain text but with a small formatting done by format_docstring, which preserves white space characters

Right, good catch!

```suggestion
        'value': escape_markdown(value)
```

should do it.

Reply:
Can `escape_markdown` break Sphinx syntax? If it does, I'd suggest to add a kwarg to that function so that when `escape=False`, it only preserves white space characters.

Reply:
I'm with you here, I came to the same conclusion but I was thinking about splitting it into `escape_plaintext` and `escape_markdown` which will call the former. I should find some time tomorrow to push the changes.

Reply:
Better yet! I like that idea, so please follow your own suggestion and disregard mine.